### PR TITLE
Remove the misguided attempt at TORCH_STATIC.

### LINKED ISSUE #20742: Libtorch C++ Silently Fails to Generate Static Library libtorch.a
## 🐛 Bug

Below is all the library generated which does not contain the libtorch.a. No error messages are observed.

libprotobuf-lite.a
libclog.a
libcpuinfo_internals.a
libcpuinfo.a
libpthreadpool.a
libqnnpack.a
libnnpack_reference_layers.a
libonnxifi_loader.a
libonnxifi_dummy.so
libfoxi_loader.a
libonnxifi.so
libfoxi_dummy.so
libfoxi.so
libgloo.a
libprotobuf.a
libgloo_builder.a
libprotoc.a
libonnx_proto.a
libc10.a
libonnx.a
libCaffe2_perfkernels_avx2.a
libCaffe2_perfkernels_avx.a
libcaffe2_protos.a
libnnpack.a
libmkldnn.a
libcaffe2.a
libTHD.a
libc10d.a
libtorch.so
libcaffe2_detectron_ops.so

## To Reproduce

I follow offical steps to clone the lastest code
```
branch master
last commit 7b9ee598d64bd35cad9afea8276ebd1c069620d9
Author: Jongsoo Park <jongsoo@fb.com>
```

Configure environment variables

```
export NO_CUDA=1  //do not need GPU here
export BUILD_SHARED_LIBS=OFF //this is instructed here https://github.com/pytorch/pytorch/blob/master/docs/libtorch.rst
export BUILD_TEST=False //I set this, b/c test binary is not needed here
```

Command to build the library
```
mkdir build && cd build
python ../tools/build_libtorch.py
```

Ensure that dependencies are installed 
`conda install numpy ninja pyyaml mkl mkl-include setuptools cmake cffi typing`

## Expected behavior
libtorch.a is generated

## Environment

Collecting environment information...
PyTorch version: 1.1.0a0+7b9ee59
Is debug build: No
CUDA used to build PyTorch: Could not collect

OS:  Red Hat
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-5)
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: No
CUDA runtime version: Could not collect

Nvidia driver version: 410.104
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.15.0
[pip3] numpydoc==0.7.0
[pip3] torch==1.1.0
[pip3] torchvision==0.2.2.post3
[conda] mkl                       2019.3                      199  
[conda] mkl-include               2019.3                      199  
[conda] mkl-service               1.1.2                    py36_3  
[conda] mkl_fft                   1.0.1            py36h3010b51_0  
[conda] mkl_random                1.0.1            py36h629b387_0  
[conda] scikit-learn              0.19.1          py36_nomklh6cfcb94_0  
[conda] scipy                     1.1.0           py36_nomklh9c1e066_0  
[conda] torch                     1.1.0a0+7b9ee59           <pip>
[conda] torch                     1.1.0                     <pip>
[conda] torchvision               0.2.2.post3               <pip>

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21704 Remove the misguided attempt at TORCH_STATIC.**

TORCH_STATIC does something, but it's not what you want.  What the
TORCH_STATIC toggle does is cause libtorch to be generated as a static
library, rather than a dynamic library.  This sounds useful, until you
realize that it doesn't affect any of the *other* libraries we also
define in PyTorch, e.g., libcaffe2 or libc10, etc. etc.  So you end up
with one library that is static... that links to a pile of dynamic
libraries.  Not useful.

The proper way to build everything "static" is to set -DBUILD_SHARED_LIBS=OFF
which will turn the interpretation of any add_library() specifier to
generate a static library (assuming you didn't override it with SHARED
which is what happened here).

There is some attempt at building static libraries in builder, but it
doesn't seem to work, and in any case we don't seem to publish them.

Fixes #20742 (but it's not enough to make static linking work)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Review by soumith — APPROVED** [review]

**Review by kostmo — CHANGES_REQUESTED** [review]
Can you hold off on this?  I'm trying to land the last piece of my unification work, and this will conflict.

**Comment by xsacha** [comment]
Essentially everything should be static or shared.
So TORCH_STATIC shouldn't exist but it should instead comply to BUILD_SHARED_LIBS CMake varaible.

Thanks for this, it has been such a simple but annoying issue.
I hope to also see static nightly/stable builds if possible?

**Comment by kostmo** [comment]
#20774 is landed, so it's ok to proceed with this now, thanks.

**Review by kostmo — APPROVED** [review]

**Review by peterjc123 — APPROVED** [review]
I'm just wondering why merging `torch` and `torch_gpu` doesn't trigger the error on the `build_shared_libs=off` builds on Windows. They are almost close to the maximum size (2GB) before merging.

**Comment by peterjc123** [comment]
@xsacha I was working on that, but still has some problems generating the new Windows AMI. I could not tell whether the issue is actually on MSBuild or the CMake side, but using this combination will throw some weird build errors, mainly related to some racing conditions when running the custom build jobs.

File: caffe2/CMakeLists.txt
Comment (ezyang) [reply caffe2/CMakeLists.txt]: I think @kostmo is the one who would know. But I think we just need to swing over all the source files from caffe2 to torch.

**Comment by ezyang** [comment]
@xsacha At the end of this I too would like to see a static libtorch build. But there are a lot of roadblocks along the way. I'm working on a write up right now.

**Comment by ezyang** [comment]
It looks after @kostmo's patch `TORCH_STATIC` is gone anyway, so this PR is moot.

**Comment by rishikhan** [comment]
@ezyang Is there any progress on the roadblock to a static build? I'd like to see this happen too.

**Comment by ezyang** [comment]
Take a look at https://github.com/pytorch/pytorch/issues/21737 at this
point the issue just needs hands.

Excerpts from Rishi Khan's message of 2019-07-12 11:25:01 -0700:
> @ezyang Is there any progress on the roadblock to a static build? I'd like to see this happen too.
>

**Pull request closed without merging**
